### PR TITLE
Fixed #30916 -- Added support for functional unique constraints.

### PR DESCRIPTION
**Pull request by hannseman**
Ticket: https://code.djangoproject.com/ticket/30916

### CODE REVIEW

File: django/db/models/constraints.py [hunk 1]
```diff
@@ -132,37 +157,57 @@ def _get_condition_sql(self, model, schema_editor):
...
-            include=include, opclasses=self.opclasses,
+            include=include, opclasses=self.opclasses, expressions=expressions,
         )
 
     def __repr__(self):
-        return '<%s: fields=%r name=%r%s%s%s%s>' % (
-            self.__class__.__name__, self.fields, self.name,
+        return '<%s: %sname=%r%s%s%s%s%s>' % (
+            self.__class__.__name__,
+            '' if not self.fields else 'fields=%s ' % repr(self.fields),
```
Comment (Ian-Foote) [inline django/db/models/constraints.py:206]: Maybe:
```suggestion
            '' if not self.fields else 'fields=%r ' % self.fields,
```

File: tests/constraints/tests.py [hunk 2]
```diff
@@ -216,6 +218,14 @@ def test_eq_with_opclasses(self):
         self.assertEqual(constraint_1, constraint_1)
         self.assertNotEqual(constraint_1, constraint_2)
 
+    def test_eq_with_expressions(self):
+        constraint = models.UniqueConstraint(Lower('title'), F('author'), name='book_func_unq')
+        same_constraint = models.UniqueConstraint(Lower('title'), 'author', name='book_func_unq')
+        another_constraint = models.UniqueConstraint(Lower('title'), name='book_func_unq')
+        self.assertEqual(constraint, same_constraint)
+        self.assertEqual(constraint, mock.ANY)
```
Comment (Ian-Foote) [inline tests/constraints/tests.py:226]: What is the point of this assertion?

File: tests/invalid_models_tests/test_models.py [hunk 3]
```diff
@@ -618,6 +618,36 @@ class Meta:
...
+
+    def test_expression_unique_constraint_required_db_features(self):
+        class Model(models.Model):
+            name = models.CharField(max_length=10)
+
+            class Meta:
+                constraints = [models.UniqueConstraint(Lower('name'), name='lower_name_unq')]
+                required_db_features = {'supports_expression_indexes'}
+
+        self.assertEqual(Model.check(databases=self.databases), [])
```
Comment (Ian-Foote) [inline tests/invalid_models_tests/test_models.py:649]: Doesn't this assertion fail if the database doesn't support expression indexes?

File: tests/invalid_models_tests/test_models.py [hunk 4]
```diff
@@ -2178,3 +2208,102 @@ class Meta:
...
+        class Bar(models.Model):
+            foo_1 = models.ForeignKey(Foo, models.CASCADE, related_name='bar_1')
+            foo_2 = models.ForeignKey(Foo, models.CASCADE, related_name='bar_2')
+
+            class Meta:
+                constraints = [
+                    models.UniqueConstraint(Lower('foo_1_id'), Lower('foo_2'), name='name'),
+                ]
+
+        self.assertEqual(Bar.check(databases=self.databases), [])
```
Comment (Ian-Foote) [inline tests/invalid_models_tests/test_models.py:2309]: I think this test would make a little more sense if we used a `CharField` for the primary key of `Foo`. It's not super important though.

**Review by Ian-Foote — COMMENTED** [review]
This is looking great! I've left a few minor comments and suggestions.

I'm a little surprised you didn't add any tests checking that we get appropriate errors raised when these unique constraints are violated. Is there a reason you didn't take that route?

I've also noticed that we now have two ways of defining a simple unique constraint on fields using `UniqueConstraint`:

```python
UniqueConstraint('field_one', 'field_two', name='constraint_one')
UniqueConstraint(fields=('field_one', 'field_two'), name='constraint_two')
```

If I understand correctly, these will become the same constraint at the database level, apart from the name. I can see how the way you have approached this makes the implementation simpler, but I think this extra ambiguity in the signature isn't ideal.

I see two options to resolve this:
* Move expressions support into the `fields` keyword argument:
```python
UniqueConstraint(fields=(Lower('field_one'), 'field_two'), name='constraint_three')
```
* Deprecate `fields` and encourage using `expressions` instead everywhere.

I think option 2 is nicest long term, but I'm not sure it's worth the deprecation cycle.

**Comment by hannseman** [comment]
@Ian-Foote Thanks for the review!

> I’m a little surprised you didn’t add any tests checking that we get appropriate errors raised when these unique constraints are violated. Is there a reason you didn’t take that route?

Haha good point! I’ve been so focused on functional indexes that I forgot the actual constraint part of this PR. I think it makes sense to extend the test `test_add_expression_unique_constraint`  with a `self.assertRaises(IntegrityError)`.

> I’ve also noticed that we now have two ways of defining a simple unique constraint on fields using UniqueConstraint.

Yes, this is to keep the `UniqueConstraint` signature as close to `Index` as possible. While implementing functional indexes the [consensus](https://github.com/django/django/pull/11929#pullrequestreview-400015927) was that using the `fields` argument to pass expressions would be confusing and a `*expressions` should be preferred.

I think it would make sense to investigate deprecating `fields`. At the moment we do not try to be smart about columns being passed as `*expressions` and instead treat all resolvable objects as expressions. This will cause a system check to trigger on databases not supporting indexes on expressions. Modifying `Index.contains_expressions` to ignore expressions resolving to `Col` should make it possible to deprecate `fields` in both `UniqueConstraint` and `Index`, but I’m thinking this is probably best discussed on the mailing list.

File: django/db/models/constraints.py (hunk 1 above)
Comment (hannseman) [reply django/db/models/constraints.py:206]: I tried to keep this consistent with `include` and `opclasses` which use the `repr` builtin.

File: tests/invalid_models_tests/test_models.py (hunk 3 above)
Comment (hannseman) [reply tests/invalid_models_tests/test_models.py:649]: This is actually a check that the `required_db_features` attribute works as intended for databases not supporting this feature: https://github.com/django/django/blob/4d8c9e83550a394af8f3c6ecc4b576c0e4e4152c/django/db/models/base.py#L2042-L2047

File: tests/invalid_models_tests/test_models.py (hunk 4 above)
Comment (hannseman) [reply tests/invalid_models_tests/test_models.py:2309]: I agree!

File: tests/invalid_models_tests/test_models.py (hunk 3 above)
Comment (Ian-Foote) [reply tests/invalid_models_tests/test_models.py:649]: Oh! I see! Maybe worth a comment?

File: django/db/backends/mysql/introspection.py [hunk 5]
```diff
@@ -242,6 +242,8 @@ def get_constraints(self, cursor, table_name):
         """
         cursor.execute(type_query, [table_name])
         for constraint, kind in cursor.fetchall():
+            if constraint not in constraints:
+                continue
```
Comment (felixxm) [inline django/db/backends/mysql/introspection.py:246]: This loop assumes that we have only columns constraints, so it can crash even now on databases with manually added functional constraints. I think we should avoid an extra query and fix this in advance by joining these two statements, e.g.

<details>

```diff
diff --git a/django/db/backends/mysql/introspection.py b/django/db/backends/mysql/introspection.py
index 649df5c6d9..b31baed714 100644
--- a/django/db/backends/mysql/introspection.py
+++ b/django/db/backends/mysql/introspection.py
@@ -211,20 +211,25 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
         # Get the actual constraint names and columns
         name_query = """
             SELECT kc.`constraint_name`, kc.`column_name`,
-                kc.`referenced_table_name`, kc.`referenced_column_name`
-            FROM information_schema.key_column_usage AS kc
+                kc.`referenced_table_name`, kc.`referenced_column_name`,
+                c.`constraint_type`
+            FROM
+                information_schema.key_column_usage AS kc,
+                information_schema.table_constraints AS c
             WHERE
                 kc.table_schema = DATABASE() AND
+                c.table_schema = kc.table_schema AND
+                c.constraint_name = kc.constraint_name AND
                 kc.table_name = %s
             ORDER BY kc.`ordinal_position`
         """
         cursor.execute(name_query, [table_name])
-        for constraint, column, ref_table, ref_column in cursor.fetchall():
+        for constraint, column, ref_table, ref_column, kind in cursor.fetchall():
             if constraint not in constraints:
                 constraints[constraint] = {
                     'columns': OrderedSet(),
-                    'primary_key': False,
-                    'unique': False,
+                    'primary_key': kind == 'PRIMARY KEY',
+                    'unique': kind in {'PRIMARY KEY', 'UNIQUE'},
                     'index': False,
                     'check': False,
                     'foreign_key': (ref_table, ref_column) if ref_column else None,
@@ -232,21 +237,6 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
                 if self.connection.features.supports_index_column_ordering:
                     constraints[constraint]['orders'] = []
             constraints[constraint]['columns'].add(column)
-        # Now get the constraint types
-        type_query = """
-            SELECT c.constraint_name, c.constraint_type
-            FROM information_schema.table_constraints AS c
-            WHERE
-                c.table_schema = DATABASE() AND
-                c.table_name = %s
-        """
-        cursor.execute(type_query, [table_name])
-        for constraint, kind in cursor.fetchall():
-            if kind.lower() == "primary key":
-                constraints[constraint]['primary_key'] = True
-                constraints[constraint]['unique'] = True
-            elif kind.lower() == "unique":
-                constraints[constraint]['unique'] = True
         # Add check constraints.
         if self.connection.features.can_introspect_check_constraints:
             unnamed_constraints_index = 0

```
</details>

What do you think?

Comment (hannseman) [reply django/db/backends/mysql/introspection.py]: > This loop assumes that we have only columns constraints, so it can crash even now on databases with manually added functional constraints.

Good point, didn't think of that.

>  I think we should avoid an extra query and fix this in advance by joining these two statements.

Absolutely. The method got a whole lot cleaner! Would be nice to join `information_schema.check_constraints` as well but doesn't look like it can be done in a clean way while taking `can_introspect_check_constraints` and MariaDB in consideration.

File: django/db/backends/mysql/introspection.py [hunk 6]
```diff
@@ -298,7 +288,7 @@ def get_constraints(self, cursor, table_name):
                 constraints[index] = {
                     'columns': OrderedSet(),
                     'primary_key': False,
-                    'unique': False,
```
Comment (felixxm) [inline django/db/backends/mysql/introspection.py:301]: All changes to MySQL introspection (except this line) can be treated  as a simplification of the current implementation, so I moved it to a separate PR #14000.

File: django/db/backends/sqlite3/introspection.py [hunk 7]
```diff
@@ -413,7 +413,7 @@ def get_constraints(self, cursor, table_name):
                     }
                 constraints[index]['columns'].append(column)
             # Add type and column orders for indexes
-            if constraints[index]['index'] and not constraints[index]['unique']:
+            if constraints[index]['index']:
```
Comment (felixxm) [inline django/db/backends/sqlite3/introspection.py:416]: This is a separate issue. I've created a new ticket-32453, added tests, and PR #14014.

File: django/db/backends/mysql/introspection.py [hunk 8]
```diff
@@ -289,7 +289,7 @@ def get_constraints(self, cursor, table_name):
...
+                    'unique': not non_unique,
```
Comment (felixxm) [inline django/db/backends/mysql/introspection.py:292]: This is also a separate issue, but it's not testable without functional constraints so we can leave it here.

File: django/db/backends/sqlite3/introspection.py (hunk 7 above)
Comment (hannseman) [reply django/db/backends/sqlite3/introspection.py:416]: Thanks, I'll rebase after it's merged.

File: docs/ref/models/constraints.txt [hunk 9]
```diff
@@ -69,10 +69,31 @@ constraint.
...
+
+    UniqueConstraint(Lower('name').desc(), 'category', name='unique_name_category')
+
+creates a unique constraint on the lowercased value of the ``name`` field in descending
+order and the ``category`` field in the default ascending order.
+
+Functional unique constraints have the same database restrictions as
+:attr:`Index.expressions`.
+
+
```
Comment (felixxm) [inline docs/ref/models/constraints.txt:96]: Chop blank line.

File: docs/ref/models/constraints.txt [hunk 10]
```diff
@@ -69,10 +69,31 @@ constraint.
 ``UniqueConstraint``
 ====================
 
-.. class:: UniqueConstraint(*, fields, name, condition=None, deferrable=None, include=None, opclasses=())
+.. class:: UniqueConstraint(*expressions, fields, name, condition=None, deferrable=None, include=None, opclasses=())
```
Comment (felixxm) [inline docs/ref/models/constraints.txt:72]: Signature is outdated. We should also add `versionchanged` annotation and a [small release notes](https://docs.djangoproject.com/en/dev/releases/4.0/#miscellaneous) about signature changes.

File: django/db/models/constraints.py [hunk 11]
```diff
@@ -79,16 +85,18 @@ class Deferrable(Enum):
 class UniqueConstraint(BaseConstraint):
     def __init__(
         self,
-        *,
-        fields,
```
Comment (felixxm) [inline django/db/models/constraints.py:83]: We cannot change the signature that much, it's backward incompatible. IMO it's acceptable to make `fields` and `name` arguments optional:
```python
    def __init__(
        self,
        *expressions,
        fields=(),
        name=None,
        ...
    ):
        if not name:
            raise ValueError('An unique constraint must be named.')
        ```
```

File: django/db/models/constraints.py [hunk 12]
```diff
@@ -132,37 +157,57 @@ def _get_condition_sql(self, model, schema_editor):
...
     def __repr__(self):
-        return '<%s: fields=%r name=%r%s%s%s%s>' % (
-            self.__class__.__name__, self.fields, self.name,
+        return '<%s: %sname=%r%s%s%s%s%s>' % (
+            self.__class__.__name__,
+            '' if not self.fields else 'fields=%s ' % repr(self.fields),
+            self.name,
+            '' if not self.expressions else ' expressions=%s' % ', '.join([
+                str(expression) for expression in self.expressions
+            ]),
```
Comment (felixxm) [inline django/db/models/constraints.py:210]: We could  keep the same format as in indexes:
```suggestion
        return '<%s: %sname=%r%s%s%s%s%s>' % (
            self.__class__.__name__,
            '' if not self.fields else "fields='%s' " % ', '.join(self.fields),
            self.name,
            '' if not self.expressions else " expressions='%s'" % ', '.join([
                str(expression) for expression in self.expressions
            ]),
```

**Review by felixxm — COMMENTED** [review]
@hannseman Thanks for this patch :+1: I left comments and pushed small edits. Please push updates in new commits.

**Comment by felixxm** [comment]
Some tests crash on Oracle, I will check them ...

**Comment by felixxm** [comment]
> Some tests crash on Oracle, I will check them ...

I fixed introspection of index uniqueness on Oracle.

File: django/db/models/constraints.py (hunk 12 above)
Comment (hannseman) [reply django/db/models/constraints.py:215]: Yeah I've been meaning to open up a PR with that. Do you think we should change the other arguments like `condition` and `opclasses` while we're at it or do we want it in a separate commit?

File: django/db/models/constraints.py (hunk 11 above)
Comment (hannseman) [reply django/db/models/constraints.py:83]: Yikes, I was a bit too fast there. Thanks.

**Comment by hannseman** [comment]
Thanks for the review. 

I don't really know how and when I managed to rebase the "Fixed #30916 -- Added support for functional constraints." commit into "Fixed introspection of unique indexes on MySQL."..

> Some tests crash on Oracle, I will check them ...

I should probably have another shot at getting Oracle running locally..

**Comment by felixxm** [comment]
> I don't really know how or when I managed to rebase the "Fixed #30916 -- Added support for functional constraints." commit into "Fixed introspection of unique indexes on MySQL."..

It's my fault, but don't worry I will change a commit message with the final rebase.

File: django/db/models/constraints.py (hunk 12 above)
Comment (felixxm) [reply django/db/models/constraints.py]: Yes a separate PR with unification sounds good.

**Comment by hannseman** [comment]
I've rebased in the changes from https://github.com/django/django/pull/14019. I opted for a rebase rather than a merge commit and reverts / updates. Let me know if you want that instead, it was just a lot easier with rebase and I only touched 01b74b5 and 7a28898.

File: docs/releases/4.0.txt [hunk 13]
```diff
@@ -275,6 +278,11 @@ Miscellaneous
   :setting:`STATIC_URL`, the leading slash is removed from that setting (now
   ``'static/'``) in the default :djadmin:`startproject` template.
 
+*  To allow expressions to be passed as positional arguments on
+   :class:`UniqueConstraint() <django.db.models.UniqueConstraint>` the
+   ``fields`` argument now default to an empty tuple and the ``name`` argument
+   default to ``None``.
```
Comment (felixxm) [inline docs/releases/4.0.txt:284]: I've decided that this change is so minor, that's not worth documenting.

**Review by felixxm — COMMENTED** [review]
@hannseman Thanks :+1: I squashed previous commits and pushed final edits to docs and schema tests.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
@hannseman Thanks again :star: :fireworks:

**Comment by hannseman** [comment]
@felixxm @Ian-Foote many thanks for the reviews ❤️

**Comment by pauloxnet** [comment]
Another great contribution. Thanks @hannseman and @felixxm
